Better enum case stemming

dpassage opened the pull request:
This fixes #138.

I tried to match the style of the files I was editing, which were different. Also, I only check for identifiers starting with `[0-9]` and literal `_`; if the new identifier ends up being a reserved word, other code in the generator already quotes it with backticks.

Thanks for putting as much work and thought into this as you do!

allevato commented on Sources/protoc-gen-swift/StringUtils.swift:
```diff
@@ -274,3 +274,12 @@ func stringToEscapedStringLiteral(_ s: String) -> String {
   return out + "\""
 }
+
+func isValidSwiftIdentifier(_ s: String) -> Bool {
```
The name of this function is a bit misleading; the comment makes it clearer why this works, but I would fear that future readers/users might think this function does what the name says in *all* cases. What about `isProtoIdentifierValidSwiftIdentifier`, which captures the notion that the input is assumed to already be a valid protobuf identifier?

tbkka replied:
I would like to use `isValidSwiftIdentifier` throughout the generator.  How hard would it be to expand it to actually do what it says?

The full spec is not small, but it's not inherently complicated, either.  It boils down to "first character must be this and the rest must be that":

 https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/zzSummaryOfTheGrammar.html

allevato replied:
If you go this route, consider precompiling a (rather large) [`NSRegularExpression`](https://developer.apple.com/reference/foundation/nsregularexpression) as a private global and matching against that. It looks like it handles the higher Unicode code points the way it should with `\U`, so the hard part is just getting everything written down in the expression.

dpassage replied:
Working on that point, thanks!

dpassage commented:
OK! `StringUtils` now has 3 validation functions:

* `isValidSwiftIdentifier`, which matches the full production from the Swift Programming Language for identifiers
* `isUsableSwiftIdentifier`, which excludes things we can't use in declarations, i.e., `_` and `$0`-`$9`
* `isValidProtobufIdentifier`, which matches valid identifiers for `.proto` files.

In the process, I added unit tests against `protoc-gen-swift` to cover these. I did run `swift package generate-xcodeproj`, but the diff was nuts - are there options you used in generating the one in the project?

In any case, hopefully this addresses the concerns.

dpassage commented:
@allevato @tbkka Can you take another look at this PR? Thanks!

kientzle commented:
The tests look yummy.

Do you have any way to test this on Linux?  I'm not sure that NSRegularExpression exists there.  If not, this may have to be coded in a somewhat more explicit fashion, building on basic functions like:
```
func isValidSwiftIdentifierCharacter(_ c: Character) -> Bool {
  switch c {
  case "a"..."z", "A"..."Z", "_": return true
  case "\u00a8"..."\u00aa": return true
  /* ... etc ... */
  default: return false
  }
}

func isValidSwiftIdentifier(_ s: String) -> Bool {
  for c in s.characters {
    if !isValidSwiftIdentifierCharacter(c) {
      return false
    }
  }
  return true
}
```

dpassage commented:
I don't have an immediate way to run this on Linux, no. If someone could help with validating my PR there I'd appreciate it, otherwise I can get it going...

dpassage commented:
Sure 'nuff, no `NSRegularExpression` on linux, working on it...

dpassage commented:
OK, so I've updated the `StringUtils` so that it compiles on Linux. I also updated some of the tests so they `import Foundation`, as `Data` isn't in the standard library.

In any case, I'm at the point in this where I'm doing more work to make things work on Linux than to address this bug in the `swift-protobuf` compiler. I'm happy to help with that effort; is that something that needs to happen in order to merge this bug fix?

tbkka commented:
I'm working through some other issues right now, but should have time to help with this later this week.

dpassage commented:
Great, thanks!

tbkka commented:
Now that #178 has been merged, you should be able to simplify this a lot.  Thanks for your patience while we work this out.

dpassage commented:
No, please, thank you!

dpassage commented:
Closing in favor of a new branch...
